### PR TITLE
Message search bar

### LINKED ISSUE #134: Message Search Bar
You're almost done with the message page! Your task is to make the search bar functional and begin implementing the dropdown (a pretty short sprint this week).

- [x] Make the search bar functional so that pressing the "Enter" button or the search icon only fetches and displays conversations in the "Unread Messages" and "All Messages" sections that contain messages with the search term(s). Every term must match (capitalization does not matter). For example, a conversation with "hi bob" should be displayed if the search terms are "Hi Bob", but a conversation with just "bob" should not be displayed.
Image

Next, begin implementing the dropdown in the messages page.

> When you click the ellipse in the top right corner of the chat, a dropdown opens up.

**Psychiatrist side:**
Screenshot 2024-04-17 at 11 18 40 PM

**Patient side:**
Screenshot 2024-04-17 at 11 18 11 PM

- [x] Currently, the dropdown in `/chatapp/ChatArea.tsx` looks the same for both patient and psychiatrist accounts. We want them to look like the above, ensuring the dropdown looks correct for each user type. Make sure to remove the "Book Appointment" option for both of them.

- [x] Link the "View Profile" button for the patient's dropdown to the psychiatrist's profile page. The pathname should be /patient/[patient_uid]/prof_profile with a query on the psych_id. This should be the exact same as the `handleGoToProfProfile` function in `frontend/src/components/psychiatrists/PsychiatristList.tsx`.

### LINKED ISSUE #135: Finalize the Messages Page
Your task is to finish implementing the dropdown in the messages page.

> When you click the ellipse in the top right corner of the chat, a dropdown opens up.

**Psychiatrist side:**
Image

**Patient side:**
Image

- [x] If a patient clicks "Mark as Unread" in the dropdown, `messagesUnreadByPatient` should be set to 1. . If a psychiatrist clicks "Mark as Unread" in the dropdown, `messagesUnreadByPsych` should be set to 1. This is just to mark some message as unread in the conversation (1 is arbitrary). Make sure the conversation moves to the "Unread Messages" section when this is clicked.
- [x] Create the following modal, so that clicking "Delete Message Thread" in the dropdown should open it up. Use `/src/assets/cancel.svg` for the cancel button. In `frontend/src/assets`, create a file called `delete.svg`. In the OKB Figma, get the SVG code for the delete button and add it to `delete.svg`.
Screenshot 2024-03-30 at 4 23 26 AM

- [x] If a patient clicks the "Delete" button in the modal, `deletedByPatient` should be set to True for the corresponding conversation, and the conversation should be removed from the patient's view but should still be visible on the psychiatrist's end as long as the psychiatrist hasn't deleted the thread.

- [x] If both `deletedByPatient` and `deletedByPsych` are set to True, then we delete the conversation from the `Conversations` collection, as well as all documents in the `Chats` collection where the `uid` and `recipientId` are the participants of the deleted conversation thread. This ensures you delete the entire message thread properly.

### LINKED ISSUE #146: Messages Bug Fixes
Your final ticket is to fix any bugs/issues with messages. Here is a list of bugs and issues that I found, but if you find any that are not mentioned here, please fix them as well! **The ticket is due on April 24 @ 10 PM and revisions are due on April 26 @ 5 PM.** These are hard deadlines as we need to have this ticket fully completed by Friday so that I can prepare the deployment for showcase and make sure everything is bug-free! As always, let me know if you guys have any questions. We are so close to the finish line! :)

I have not merged your branch to `main` yet due to these bugs, so continue working on your `message_search_bar` branch.

- [x] Task 1: Fix Bug 1

### Bug 1: Incomplete Deletion of Message Threads
Currently, when a user selects "Delete Message Thread" for a conversation, only the conversation disappears from the list on the left, while the messages remain visible in the chat area. This inconsistency reveals a flaw in the current approach. To address this, I recommend adding `deletedByPatient` and `deletedByPsych` fields not only to conversation documents in the `Conversations` collection but also to individual message documents in the `Chats` collection.

Here's how it would work: The `deletedByPatient` and `deletedByPsych` fields for a message will determine whether that particular message should be displayed to the respective user. Meanwhile, the `deletedByPatient` and `deletedByPsych` fields for a conversation document will control whether the conversation appears in the sidebar and whether it should be removed from the collection. For example, if the patient sends a new message in that conversation, `deletedByPatient` should be set to `False` if it was previously set to `True` so that the conversation can appear in their sidebar. I can clarify further if this is confusing, so please let me know.

This change ensures that when a user (patient or psychiatrist) deletes a message thread, the action reflects on each message within that thread rather than just the conversation itself so that the chat area is properly cleared out. Additionally, this approach resolves another issue: if a patient deletes a message thread and later sends a new message to the same psychiatrist, this new conversation does not appear in the patient's sidebar. This is because the `deletedByPatient` status is still set to `True`. By marking individual messages as deleted, the visibility of messages for a user will directly correlate with the deletion status on each message for that user.

- [x] Fix Bug 2

### Bug 2: Chat Area Displays Incorrectly When No Conversation is Selected
**Description:** When the "Messages" tab in the Navbar is clicked and no conversation has been selected, the chat area should be empty, displaying neither the name area nor the message composer. However, the chat area is not displaying as expected; elements that should be hidden (such as the name area and message composer) are still visible.

**Steps to Reproduce This Bug:**
1. Log in as either a patient or a psychiatrist.
2. Click on the "Messages" tab in the Navbar without selecting any conversation.
3. Observe the chat area.

**Expected Result:**
The chat area should be completely empty with no elements visible (no name area or message composer), indicating no conversation has been selected.

- [x] Task 3: Add ellipsis to messages displayed in the sidebar that are really long. Messages should be displayed at at most 3 lines long like the image below.

Image

- [x] Task 4: When a patient attempts to message a psychiatrist for the first time, that chat should say "Start a conversation with Gloria Shi." Refer to the image below.

Image

- [x] Task 5: Make the message composer match the Figma design. For instance, this will involve making the color of the "Send" button the OKB Blue color in `colors.ts`.

Current Design:
Image

Desired Design:
Image

### PR DESCRIPTION
**Pull request by leaneying**
## Overview

For this ticket, we made the search bar functional so that pressing the "Enter" button or the search icon only fetches and displays conversations in the "Unread Messages" and "All Messages" sections that contain messages with the search term (ignoring capitalization). We also changed the dropdown menu so that it changed depending on whether the current user was a patient or psychiatrist to reflect the capabilities of their role. Finally, we linked the "View Profile" button for the patient's dropdown to the psychiatrist's profile page.

## Replication

To test the code, one can clone the repository, cd frontend, and yarn dev the terminal. After accessing the localhost, if you have not already made an account, sign up, then login and navigate to messages in the top right of the navbar. If you have no messages, create at least one more account (psychiatrist if your account is patient and vice versa) to send yourself messages. You will be able to see how searching in the search bar will give you the conversations containing that search term. Additionally, you will also see that pressing the drop down menu will provide you with a different list depending on your role and that pressing the “view profile” button in the dropdown menu for patients will take you to that psychiatrist’s profile. 

## Changes Made

Updated ConversationList.tsx to include a prop for SearchInput, so that conversation lists only display conversations where SearchInput is equal to the other user’s name. If SearchInput is empty, we still return all conversations as normal. 

We updated ChatArea.tsx to also take in a role prop that is determined by the user’s id and whether or not fetching professionalData returns null. We also added UseEffect functionality to determine the psych id and patient id of the two people in the conversation. We turned the “View Profile” list item into a button, and used these two id’s to correctly route the button to the psychiatrist’s profile. 

## Test Coverage

We tested this feature by searching messages in the search bar, making sure to change capitalization and also change whether it was a patient or psychiatrist sending it. We also checked for patients and psychiatrists that pressing the drop down menu changed depending on user role, and that pressing “View Profile” would route you to the correct psychiatrist’s profile.

## Next Steps 

For our next PR, we will finish implementing the dropdown in the messages page by adding functionality to the “Mark as Unread” and “Delete” buttons. 

## Related PRs or Issues

This is related to our PR and Issue #117 which involved making the sidebar functional and implementing real-time messaging, as well as PR and Issue #128 that finishedthe implementation of the sidebar, ensuring that the "Unread Messages" and "All Messages" sections work properly.

## Screenshots

<details>

The dropdown menu on the patient side is correct:
temp1

The dropdown menu on the psychiatrist side is correct:
temp2

Pressing “View Profile” routes the user to the correct psychiatrist’s profile:
temp3

</details>

## Overview

For this ticket, we finished implementing the dropdown in the messages page. This involved making both “Mark as Unread” and “Delete Message Thread” functional.

## Replication

To test the code, one can clone the repository, cd frontend, and yarn dev the terminal. After accessing the localhost, if you have not already made an account, sign up, then login and navigate to messages in the top right of the navbar. If you have no messages, create at least one more account (psychiatrist if your account is patient and vice versa) to send yourself messages. You will be able to see that pressing the drop down menu will provide you with a different list depending on your role and that pressing the “Mark as Unread” will mark the conversation as unread and move it into the unread section until you click on it to read it. Additionally, pressing “Delete Message Thread” will make a modal pop up to confirm whether you want to delete the thread. If you press cancel, the modal will disappear and the conversation will not be deleted. If you press delete, then the conversation will be deleted.

## Changes Made

We imported delete.svg from the figma for the modal. We also created a new component DeleteModal.tsx to create the pop-up and make it functional in terms of calling function by pressing the buttons, and making it conditionally visible. We updated ChatArea.tsx to call DeleteModal. We created and implemented functions openDeleteModal, closeDeleteModal, and handleDelete to create the functionality of DeleteModal. In DeleteModal, we queried based on role and deleted the document in FireBase (making sure to update either deletedByPatient or deletedByPsych as necessary)

## Test Coverage

We tested this feature by going to different conversations, testing marking them as unread and making sure that they went to the “unread” section, and then testing deleting them, making sure that their conversation was deleted in the sidebar and that in Firebase, it updated to say that either deletedByPatient or deletedByPsych was set to true.

## Related PRs or Issues

This is related to our PR and Issue #117 which involved making the sidebar functional and implementing real-time messaging, as well as PR and Issue #128 that finished the implementation of the sidebar, ensuring that the "Unread Messages" and "All Messages" sections work properly, as well as PR and Issue #134 where we made the search bar functional and began implementing the dropdown menu.

## Screenshots

<details>

Before pressing anything, you see that the conversation is set to read
temp1

After pressing “mark as unread”, the conversation moves to “unread”
temp2

After clicking on the conversation again to read it and then pressing “delete this message thread”, you get the pop-up
temp3

After confirming your choice to delete the thread, the conversation disappears from the sidebar
temp4

## Overview

For our final ticket we fixed the bugs. This included the bug involving incomplete deletion of message threads (including the conversation not reappearing after messaging a previously-deleted conversation), chat area incorrectly displaying when no messages are selected, and small fixes such as the recent message in the sidebar only showing the first 3 lines of a message (dynamically adjusts) with an ellipse at the end, adding text saying “start a conversation with …” when starting a new conversation, and making the send button the correct color. 

## Replication

To test the code, one can clone the repository, cd frontend, and yarn dev the terminal. After accessing the localhost, if you have not already made an account, sign up, then login and navigate to messages in the top right of the navbar. If you have no messages, create at least one more account (psychiatrist if your account is patient and vice versa) to send yourself messages. You can delete a conversation and notice that it will completely disappear from both the chat area and the sidebar. When re-messaging that person, you will see that it correctly reappears in the sidebar and the correct text “start a conversation with …” will appear in the chat area and no previous messages. If you send a long message, only the first three will appear in the sidebar preview with an ellipse at the end and you can change the window size and you will see that this is dynamic and will always be three lines.

## Changes Made
In ConversationItem, we adjusted the css so that it would only display the first 3 lines and any overflow would be represented by ellipses. In ChatArea, we adjusted the handleDelete function to query and delete from both the conversations and chats to correctly delete both. We implemented this by adding the deletedByPatient and deletedByPsych fields that were already in Conversations to Chats. In MessageList, we created a filteredMessages function to filter individual messages by deletion status as well as user role. In MessageComposer, we made sure to update deletedByPatient and deletedByPsych and reset them when needed. Finally, we HTML in MessageList for the “start a conversation with …” message if there were no messages after filtering and in MessageComposer making the send button OKB blue. 

## Test Coverage
We tested these features by creating, deleting, and recreating conversations and making sure everything was deleting and reappearing as they should with the new conversation method appearing as it should. We also tested by sending long messages and making sure they were formatted correctly in the sidebar and were dynamic with changing the browser window size. 
## Related PRs or Issues

This is related to our PR and Issue #117 which involved making the sidebar functional and implementing real-time messaging, as well as PR and Issue #128 that finished the implementation of the sidebar, ensuring that the "Unread Messages" and "All Messages" sections work properly, as well as PR and Issue #134 where we made the search bar functional and began implementing the dropdown menu as well as PR and Issue #135 “Finalize the Messages Page”.

## Screenshots

<details>

When starting a new conversation, the new conversation message appears at the top of the chat area (additionally, the send button is the correct button)
temp1

When deleting a conversation, it disappears from both the sidebar and the messages disappear from the chat area
temp2

When re-messaging in a previously deleted conversation, it will properly appear in the sidebar and chat area with the new conversation message and no previous messages
temp3

When sending a long message, it will only show the first three lines (and an ellipse) in the sidebar and it is dynamic. 
temp4
temp5

### CODE REVIEW

**Review by amywu4 — CHANGES_REQUESTED** [review]
All the completed tasks for the first sprint in this PR (Message Search Bar) look good! 

However, there are two bugs in the second sprint.
### Bug 1: Incorrect Behavior in Marking Conversations as Unread for Psychiatrists (Works fine on the Patient side)
**Description:**
There is a bug affecting the "Unread Messages" functionality for psychiatrists. Specifically, after marking a conversation as unread and following a series of navigation steps, the conversation does not disappear from the "Unread Messages" section. I suspect this issue might be related to incorrect routing parameters used when loading pages as a psychiatrist. Either way, please fix both issues.

**Steps to Reproduce This Bug:**
1. Log in as a psychiatrist.
2. Navigate to the "Messages" section by clicking the "Messages" button in the Navbar.
3. Select a conversation.
4. Check the URL of the conversation. It incorrectly shows `/patient/[patient_id]/messages?psych_id=[psych_id]&psych_name=[patient's first name]%20[patient's last name]` (bug).
5. Mark the selected conversation as unread.
6. Reload the page. This erroneously redirects to the psychiatrist dashboard instead of staying on the messages page which is most probably due to the incorrect path stated in Step 4. Fixing Step 4 should fix this. (bug)
7. Click on the "Messages" button in the Navbar again.
8. Click on the previously marked unread conversation.
9. Observe that the conversation appears in the "All Messages" section but remains in the "Unread Messages" section (bug).

**Expected Result:**
Step 4: The URL should correctly represent the psychiatrist's perspective, expected to be `/psychiatrist/[psych_id]/messages?patient_id=[patient_id]&patient_name=[patient's first name]%20[patient's last name]`.
Step 6: Reloading the messages page should stay on the messages page.
Step 9: After marking a conversation as unread and then accessing it again by following the steps above, it should disappear from the "Unread Messages" section.

**Actual Result:**
The conversation does not disappear from the "Unread Messages" section after being accessed post marking it unread. Moreover, the URL is incorrect, reflecting patient parameters instead of psychiatrist parameters.


### Bug 2: Incomplete Deletion of Message Threads
Currently, when a user selects "Delete Message Thread" for a conversation, only the conversation disappears from the list on the left, while the messages remain visible in the chat area. This inconsistency reveals a flaw in the current approach. To address this, I recommend adding `deletedByPatient` and `deletedByPsych` fields not only to conversation documents in the `Conversations` collection but also to individual message documents in the `Chats` collection.

Here's how it would work: The `deletedByPatient` and `deletedByPsych` fields for a message will determine whether that particular message should be displayed to the respective user. Meanwhile, the `deletedByPatient` and `deletedByPsych` fields for a conversation document will control whether the conversation appears in the sidebar and whether it should be removed from the collection. For example, if the patient sends a new message in that conversation, `deletedByPatient` should be set to `False` if it was previously set to `True` so that the conversation can appear in their sidebar. I can clarify further if this is confusing, so please let me know.

This change ensures that when a user (patient or psychiatrist) deletes a message thread, the action reflects on each message within that thread rather than just the conversation itself so that the chat area is properly cleared out. Additionally, this approach resolves another issue: if a patient deletes a message thread and later sends a new message to the same psychiatrist, this new conversation does not appear in the patient's sidebar. This is because the `deletedByPatient` status is still set to `True`. By marking individual messages as deleted, the visibility of messages for a user will directly correlate with the deletion status on each message for that user.

File: frontend/src/components/chatapp/MessageList.tsx [hunk 1]
```diff
@@ -90,20 +66,32 @@ const MessageList: React.FC = () => {
...
-            <MessageItem message={msg} />
+
+  // Filter messages based on deletion status and role
+  const filteredMessages = messages.filter(msg => {
+    return (msg.role === 'patient' ? !msg.deletedByPatient : !msg.deletedByPsych);
+  });
+
+  if (filteredMessages.length === 0) {
+    return (
+      <div className="page-background text-center text-gray-400 pt-4 font-montserrat">Start a conversation with {psychName}</div>
```
Comment (amywu4) [inline frontend/src/components/chatapp/MessageList.tsx]: Italicize "Start a conversation with {psychName}"

**Review by amywu4 — CHANGES_REQUESTED** [review]
Nice job! I added some comments for some small adjustments that need to be made. 
### Delete Message Bug:
- When I delete a message thread as a patient, only my messages are deleted in the chat area, but the psychiatrist's messages aren't deleted. We want messages from both users to be deleted on the patient's end and the "Start a conversation with {psychName}" to appear. 
- When I delete a message thread as a psychiatrist, none of the messages displayed in the chat area are deleted. We want messages from both users to be deleted on the psychiatrist's end and the "Start a conversation with {patientName}" to appear. 
- To summarize, all messages sent from both users should be deleted in the chat area for the user who deleted the message thread; no message should be deleted on the other user's end as a result of this.

### Message Composer Bug:
- Is it possible to fix the message composer so that there aren't these 2 empty spots here? 
Screenshot 2024-04-25 at 1 25 58 AM

### Also, there are a couple of bugs that arose that needs to be fixed:
- The search bar functionality: The search bar no longer works. 
- Also, the number of unread messages is 1 no matter how many messages are sent that have not been read yet.

**Review by amywu4 — APPROVED** [review]
LGTM! :)
